Search Capability Full Built out

rohitmusti opened the pull request:
@bvkin and I have been working on the search capability to address many of the bugs that @rdebeasi was able to point out from our last PR (since closed).

**Changes**:
We added Gruntfile.js to build the search index. There was nothing too fancy here and the README has been updated to show the extra step in building it. The Gruntfile.js writes the search index to /static/js/lunr/PagesIndex.json.

The actual index is in a json format. Then we added the lunr script for searching the index at /static/js/search.js. This uses the lunrjs library for searching. Also included in this file, you'll find some html elements we used to keep the style consistent with the rest of the website. This also includes regular expressions for cleaning up the search results. I am looking for a more elegant solution within lunrjs, however there is nothing available at the moment.

I then created searchButton.js to handle the search bar capability (I figured no one wanted console searches). This is a fairly straightforward clickhandler.

Next I created the ./content/page/search.md file as a page to redirect the searches to. In order to build this page to handle the searches, I created a new partial at /layouts/page/search.html. This is a fairly custom partial for handling the searches. I also created two other partials footer-searchbutton.html and footer-search.html, as well as modifying fooder.html and /nav/menus/main.html. This was to make the search function work across the entire site.

Finally, I had to edit the config.toml file for testing purposes. On the test site at [rohitmusti.github.io/practice-library/](https://rohitmusti.github.io/practice-library), the base URL is set to rohitmusti.github.io/practice-library. I reset the URL to rht-labs.github.io/practice-library/ on the PullRequest branch so that it flawlessly integrates with the current site (or any future URL).

A complete demo of the search feature can be found at [rohitmusti.github.io/practice-library/](https://rohitmusti.github.io/practice-library)

**Note**:
The merge conflicts were from that I was using a slightly newer version of hugo. The updated version of hugo doesn't affect the website building process, it is just a little bit more stable so I went ahead and resolved those conficts.

rdebeasi commented:
Looks awesome! This works really well and is built in a thoughtful way. With a few last changes I think we'll be ready to go.

### As part of this PR
- Wrap your JavaScript code in an [IIFE](https://developer.mozilla.org/en-US/docs/Glossary/IIFE) to avoid creating global variables. (Right now there are a few globals, including one called `url`.)
  - You get this behavior "for free" when using ES6 modules, but we're just using `<script>` tags here, so we need to add the IIFE ourselves.
- Right-align the search box and add a little space between it and the nav items. You may need to reduce the size of the search box and/or the button in order to add this margin.
- Add search box styles to `custom.css` rather than using inline styles in `layouts/menus/main.html`.
- Add "install Node.js" to the "developing" section of the readme.
  - [Installers for Windows and macOS](https://nodejs.org/en/download/)
  - [Install on Linux via package manager](https://nodejs.org/en/download/package-manager/)
- In `search.js`, use `window.location.origin`, which includes the port number, instead of `window.location.hostname`. This allows search to work when doing local development.

### In a future PR
- Investigate getting article metadata from Lunr so that we don't need to use regex to generate article titles. ([Relevant xkcd](https://xkcd.com/1171/))
- Use either `let` or `var`, but not both. (`search.js` uses `let` at the top. Without a babel build step, this won't work in older browsers.)
- To simplify maintenance in the future, remove HTML strings from JavaScript files. Either retrieve HTML from elements on the page for from `<template>` tags.

rohitmusti commented:
Hey @rdebeasi 

Thank you for the feedback. I have pushed up some changes that ought to address the following issues.

**Changes**
1. I added documentation to the README linking to the Node.js installation page.
2. I wrapped my JS code in an IIFE in order to avoid the global variables problem.
3. I right-aligned the search bar and introduced some spacing in the navigation space.
4. I moved the search box styling to custom.css instead of using in-line styling
5. I swapped in window.location.origin to ease development.

rdebeasi commented:
Looks great! Just a couple of last things:

- [The search index](https://github.com/rohitmusti/practice-library/blob/233dc73aebba41846642dac746363ef008785d3e/docs/search/index.html) has some duplicated tags. Deleting the associated code from `/layouts/page/search.html` should fix this.
- The [user story mapping](https://github.com/rohitmusti/practice-library/blob/233dc73aebba41846642dac746363ef008785d3e/docs/practices/user-story-mapping/index.html) page has a conflict that wasn't fully resolved. Running `hugo` again and adding/committing the result should fix this.
